rasa/core/policies/rule_policy.py

```diff
@@ -211,15 +241,132 @@ def _create_loop_unhappy_lookup_from_states(
+        if predicted_action_name == tracker.active_loop_name:
+            rasa.core.test.emulate_form_rejection(tracker)
+            predicted_action_name = self._predict_next_action(
+                tracker, domain, interpreter
+            )
+            if not predicted_action_name or predicted_action_name == gold_action_name:
+                return
+
+        tracker_type = "rule" if tracker.is_rule_tracker else "story"
+        return (
+            f"Action '{gold_action_name}' in {tracker_type} "
+            f"'{tracker.sender_id}' was predicted incorrectly by "
+            f"the {self.__class__.__name__} as action "
+            f"'{predicted_action_name}'."
+        )
+
+    def _find_contradicting_rules(
+        self,
+        trackers: List[TrackerWithCachedStates],
+        domain: Domain,
+        interpreter: NaturalLanguageInterpreter,
+    ) -> None:
+        logger.debug("Started checking rules and stories for contradictions.")
+        self._silent_prediction = True
+        error_messages = []
+        pbar = tqdm(
+            trackers,
+            desc="Processed trackers",
+            disable=rasa.shared.utils.io.is_logging_disabled(),
+        )
+        for tracker in pbar:
+            running_tracker = tracker.init_copy()
+            running_tracker.sender_id = tracker.sender_id
+            # the first action is always unpredictable
+            next_action_is_unpredictable = True
+            for event in tracker.applied_events():
+                # do not run prediction on unpredictable actions
+                if not isinstance(event, ActionExecuted):
+                    running_tracker.update(event)
+                    continue
+
+                if event.action_name == RULE_SNIPPET_ACTION_NAME:
+                    # notify that we shouldn't check that the action after
+                    # RULE_SNIPPET_ACTION_NAME is unpredictable
+                    next_action_is_unpredictable = True
+                    # do not add RULE_SNIPPET_ACTION_NAME event
+                    continue
+
+                if next_action_is_unpredictable or event.unpredictable:
+                    next_action_is_unpredictable = False  # reset unpredictability
+                    running_tracker.update(event)
+                    continue
+
+                gold_action_name = event.action_name or event.action_text
+                error_message = self._check_prediction(
+                    running_tracker, domain, interpreter, gold_action_name
+                )
+                if error_message:
+                    error_messages.append(error_message)
+
+                running_tracker.update(event)
+
+        self._silent_prediction = False  # turn off silent logging
+        if error_messages:
+            error_messages.append(
+                "Please update your stories and rules so that "
+                "they don't contradict each other."
+            )
+            raise InvalidRule("\n".join(error_messages))
```

Comment:
If there is a newline for every error message, then we should probably format it with `- ` to make it a list?